src/Pos/Launcher/Mode.hs

```diff
@@ -0,0 +1,168 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell     #-}
+{-# LANGUAGE TypeFamilies        #-}
+
+module Pos.Launcher.Mode
+    ( InitMode
+    , runInitMode
+    , InitModeContext(..)
+    , newEmptyInitModeContext
+    , putInitRef
+    , InitRef(..)
+    ) where
+
+import           Universum
+
+import           Control.Lens          (makeLensesFor)
+import qualified Control.Monad.Reader  as Mtl
+import           Ether.Internal        (HasLens (..))
+import           Mockable.Production   (Production)
+import           System.IO.Unsafe      (unsafeInterleaveIO)
+import           System.Wlog           (HasLoggerName (..))
+
+import           Pos.Block.Core        (Block, BlockHeader)
+import           Pos.Block.Types       (Undo)
+import           Pos.Context.Context   (GenesisLeaders, GenesisUtxo, NodeParams)
+import           Pos.Core              (IsHeader)
+import           Pos.DB                (NodeDBs)
+import           Pos.DB.Block          (MonadBlockDBWrite (..), dbGetBlockReal,
+                                        dbGetBlockReal', dbGetHeaderReal,
+                                        dbGetHeaderReal', dbGetUndoReal, dbGetUndoReal',
+                                        dbPutBlundReal)
+import           Pos.DB.Class          (MonadBlockDBGeneric (..), MonadDB (..),
+                                        MonadDBRead (..))
+import           Pos.DB.Redirect       (dbDeleteReal, dbGetReal, dbPutReal,
+                                        dbWriteBatchReal)
+import           Pos.ExecMode          (ExecMode (..), ExecModeM)
+import           Pos.Lrc.Context       (LrcContext)
+import           Pos.Slotting.Class    (MonadSlots (..))
+import           Pos.Slotting.Impl.Sum (SlottingContextSum, currentTimeSlottingReal,
+                                        getCurrentSlotBlockingReal,
+                                        getCurrentSlotInaccurateReal, getCurrentSlotReal)
+import           Pos.Slotting.MemState (MonadSlotsData (..), SlottingVar,
+                                        getSlottingDataReal, getSystemStartReal,
+                                        putSlottingDataReal, waitPenultEpochEqualsReal)
+import           Pos.Ssc.Class.Helpers (SscHelpersClass)
+import           Pos.Ssc.Class.Types   (SscBlock)
+import           Pos.Util              (Some (..))
+
+data InitRef a = InitRef
+    { initRefVar :: MVar a
+    , initRefVal :: a }
+
+makeLensesFor [ ("initRefVal", "initRefValL") ] ''InitRef
+
+newInitRef :: IO (InitRef a)
+newInitRef = do
+    m <- newEmptyMVar
+    r <- unsafeInterleaveIO (readMVar m)
```

Comment:
It definitely needs some comments. Also we put haddock comment at the beginning of each module.